dask_sql/datacontainer.py

```diff
@@ -198,11 +200,10 @@ def __init__(self, func, row_udf: bool, params, return_type=None):
 
         self.names = [param[0] for param in params]
 
-        if return_type is None:
-            # These UDFs go through apply and without providing
-            # a return type, dask will attempt to guess it, and
-            # dask might be wrong.
-            raise ValueError("Return type must be provided")
+        # validate UDF metadata
+        for dt in (*(param[1] for param in params), return_type):
+            _ = python_to_sql_type(dt)
```

Comment:
Happy to push this off to a follow up PR if it would require a lot of time to implement, but since we are computing the SQL types of the params / return Python types here, it could be worthwhile to cache this information now and grab it when we later call `python_to_sql_type` for UDF preparation:

https://github.com/dask-contrib/dask-sql/blob/8a7330948a7a8603c4e25297c2038f1f7541aad2/dask_sql/context.py#L751
https://github.com/dask-contrib/dask-sql/blob/8a7330948a7a8603c4e25297c2038f1f7541aad2/dask_sql/context.py#L775

Reply:
I played around with this a bit and ended up getting something working in https://github.com/brandon-b-miller/dask-sql/pull/1 🙂

Reply:
Thanks @charlesbluca ! will take a look at this later this morning